lib/classes/Variables.test.js

```diff
@@ -40,110 +47,115 @@ describe('Variables', () => {
 
   describe('#loadVariableSyntax()', () => {
     it('should set variableSyntax', () => {
-      const serverless = new Serverless();
-
+      // eslint-disable-next-line no-template-curly-in-string
       serverless.service.provider.variableSyntax = '\\${{([ ~:a-zA-Z0-9._\'",\\-\\/\\(\\)]+?)}}';
-
       serverless.variables.loadVariableSyntax();
       expect(serverless.variables.variableSyntax).to.be.a('RegExp');
     });
   });
 
   describe('#populateService()', () => {
-    it('should call populateProperty method', () => {
-      const serverless = new Serverless();
-
-      const populatePropertyStub = sinon
-        .stub(serverless.variables, 'populateObject').resolves();
-
+    it('should call loadVariableSyntax and then populateProperty', () => {
+      const loadVariableSyntaxStub = sinon.stub(serverless.variables, 'loadVariableSyntax')
+        .returns();
+      const populateObjectStub = sinon.stub(serverless.variables, 'populateObject').resolves();
       return expect(serverless.variables.populateService()).to.be.fulfilled
-      .then(() => {
-        expect(populatePropertyStub.called).to.be.true;
-      })
-      .finally(() => serverless.variables.populateObject.restore());
+        .then(() => {
+          expect(loadVariableSyntaxStub).to.be.calledOnce;
+          expect(populateObjectStub).to.be.calledOnce;
+          expect(loadVariableSyntaxStub).to.be.calledBefore(populateObjectStub);
+        })
+        .finally(() => {
+          populateObjectStub.restore();
+          loadVariableSyntaxStub.restore();
+        });
     });
-
-    it('should use variableSyntax', () => {
-      const serverless = new Serverless();
-
-      const variableSyntax = '\\${{([ ~:a-zA-Z0-9._\'",\\-\\/\\(\\)]+?)}}';
-      const fooValue = '${clientId()}';
-      const barValue = 'test';
-
-      serverless.service.provider.variableSyntax = variableSyntax;
-
-      serverless.service.custom = {
-        var: barValue,
-      };
-
-      serverless.service.resources = {
-        foo: fooValue,
-        bar: '${{self:custom.var}}',
-      };
-
-      return serverless.variables.populateService().then(() => {
-        expect(serverless.service.provider.variableSyntax).to.equal(variableSyntax);
-        expect(serverless.service.resources.foo).to.equal(fooValue);
-        expect(serverless.service.resources.bar).to.equal(barValue);
+    it('should remove problematic attributes bofore calling populateObject with the service',
+      () => {
+        const populateObjectStub = sinon.stub(serverless.variables, 'populateObject', (val) => {
```

Comment:
Why do you delegate the unit test _into_ the stub here?

It is cleaner, if you just stub and then check _how_ the stub was called afterwards like this:
```
const populateObjectStub = sinon.stub(serverless.variables, 'populateObject').resolves();
return expect(serverless.variables.populateService()).to.be.fulfilled
.then(() => {
  expect(populateObjectStub).to.have.been.calledWithExactly(serverless.service);
  const callVar = _.first(populateObjectStub.firstCall.args); 
  expect(callVar.provider.variableSyntax).to.be.undefined;
  ...
});
```

Otherwise you melt stub execution and result checking into one blob.

Reply:
The code under test, for reference: https://github.com/erikerikson/serverless/blob/c18eb70a28d61317007091ae765585adc7cd37f6/lib/classes/Variables.js#L98
This particular test is validating temporary state that only exists from just prior to the invocation of "populateObject" and only persists until immediately after its resolution.  After populateService is settled, the state cannot be validated.